saturn_client/core.py

```diff
@@ -0,0 +1,161 @@
+    """
+
+    _options = None
+    _saturn_api_version = None
+
+    def __init__(
+        self,
+        url: Optional[str] = None,
+        api_token: Optional[str] = None,
+    ):
+        """
+        Create a ``SaturnConnection`` to interact with the API.
+
+        :param url: URL for the SaturnCloud instance.
+        :param api_token: API token for authenticating the request to Saturn API.
+            Get from `/api/user/token`
+        """
+        self.settings = Settings(url, api_token)
+
+        # test connection to raise errors early
+        self._saturn_api_version = self.saturn_api_version
+
+    @property
+    def url(self):
+        """URL of Saturn instance"""
+        return self.settings.url
+
+    @property
+    def saturn_api_version(self):
+        """Version of Saturn API"""
+        if self._saturn_api_version is None:
+            url = urljoin(self.url, "api/status")
+            response = requests.get(url, headers=self.settings.headers)
+            if not response.ok:
+                raise ValueError(response.reason)
+            self._saturn_api_version = response.json()["version"]
+        return self._saturn_api_version
+
+    @property
+    def options(self):
+        """Options for various settings"""
+        if self._options is None:
+            url = urljoin(self.url, "api/info/servers")
+            response = requests.get(url, headers=self.settings.headers)
+            if not response.ok:
+                raise ValueError(response.reason)
+            self._options = response.json()
+        return self._options
+
+    @property
+    def describe_sizes(self) -> Dict[str, str]:
+        """Instance size options"""
+        return self.options["sizes"]
+
+    def create_project(
+        self,
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+        image_uri: Optional[str] = None,
+        start_script: Optional[str] = None,
+        environment_variables: Optional[Dict] = None,
+        working_dir: Optional[str] = None,
+        workspace_settings: Optional[Dict] = None,
+    ) -> Dict[str, Any]:
+        """
+        Create a project from scratch
+
+        :param name: Name of project.
+        :param description: Short description of the project (less than 250 characters).
+        :param image_uri: Location of the image. Example:
+            485185227295.dkr.ecr.us-east-1.amazonaws.com/saturn-dask:2020.12.01.21.10
+        :param start_script: Script that runs on start up. Examples: "pip install dask"
+        :param environment_variables: Env vars expressed as a dict. The names will be
+            coerced to uppercase.
+        :param working_dir: Location to use as working directory. Example: /home/jovyan/project
+        :param workspace_settings: Setting for the jupyter associated with the project.
+            These include: `size`, `disk_space`, `auto_shutoff`, and `start_ssh`. The options
+            for these are available from `conn.options`.
```

Comment:
In client libraries like this, I have a strong preference for objects instead of dictionaries to hold information like this. So, instead of allowing this much customization by a schema-less dictionary, you'd have to create a `WorkspaceSettings` class like this

```python
@dataclass
class WorkspaceSettings:
    size: str
    disk_space: str
    auto_shutoff: str
    start_ssh: bool = False
```

Some benefits of this compared to just JSON:

* if you run old code with a new version of the library, you get an informative error immediately if a breaking schema change has been made (like a new non-optional field has been added)
* validation can be encapsulated in the class, instead of repeated in every method that reuses this type of config object

If you don't like this option, then I recommend individual keyword arguments in `create_project()` for each of these items.

I think the pattern of "important config as a free-form dictionary" is a bad one for client libraries like this project.